test/gtest/uct/ib/test_ib.cc

```diff
@@ -4,297 +4,257 @@
 
-    typedef struct {
-        unsigned length;
-        /* data follows */
-    } recv_desc_t;
+/* don't call uct_test() here, because we don't
+ * want to init UCT for all tests */
+void test_uct_ib::init() {
+    ibctx = open_device(&ibdev);
+    query_port(ibctx, &port_attr);
+}
 
-    typedef struct {
-        unsigned have_pkey; /* if 1 - means that the configured pkey was found */
-        unsigned have_lmc;  /* if 1 - means that the lmc is higher than zero */
-#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
-        unsigned have_valid_gid_idx;
-#endif
-    } ib_port_desc_t;
-
-    static ucs_status_t ib_am_handler(void *arg, void *data, size_t length,
-                                      unsigned flags) {
-        recv_desc_t *my_desc  = (recv_desc_t *) arg;
-        uint64_t *test_ib_hdr = (uint64_t *) data;
-        uint64_t *actual_data = (uint64_t *) test_ib_hdr + 1;
-        unsigned data_length  = length - sizeof(test_ib_hdr);
-
-        my_desc->length = data_length;
-        if (*test_ib_hdr == 0xbeef) {
-            memcpy(my_desc + 1, actual_data , data_length);
-        }
-        ++test_uct_ib::ib_am_handler_counter;
-        return UCS_OK;
+void test_uct_ib::cleanup() {
+    if (ibctx != NULL) {
+        ibv_close_device(ibctx);
+        ibctx = NULL;
+        ibdev = NULL;
     }
 
-    void pkey_find(const char *dev_name, unsigned port_num, struct ibv_port_attr port_attr,
-                  struct ibv_context *ibctx, ib_port_desc_t *port_desc) {
-         uint16_t table_idx, pkey, pkey_partition;
-         uct_ib_iface_config_t *ib_config = ucs_derived_of(m_iface_config, uct_ib_iface_config_t);
-
-         /* check if the configured pkey exists in the port's pkey table */
-           for (table_idx = 0; table_idx < port_attr.pkey_tbl_len; table_idx++) {
-               if(ibv_query_pkey(ibctx, port_num, table_idx, &pkey)) {
-                   UCS_TEST_ABORT("Failed to query pkey on port " << port_num << " on device: " << dev_name);
-               }
-               pkey_partition = ntohs(pkey) & UCT_IB_PKEY_PARTITION_MASK;
-               if (pkey_partition == (ib_config->pkey_value & UCT_IB_PKEY_PARTITION_MASK)) {
-                   port_desc->have_pkey = 1;
-                   break;
-               }
-           }
+    if (initialized) {
+        uct_test::cleanup();
+        initialized = false;
     }
+}
 
-#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
-    void test_eth_port(struct ibv_device *device, struct ibv_port_attr port_attr,
-                       struct ibv_context *ibctx, unsigned port_num,
-                       ib_port_desc_t *port_desc) {
-
-        union ibv_gid gid;
-        uct_ib_md_config_t *md_config = ucs_derived_of(m_md_config, uct_ib_md_config_t);
-        char md_name[UCT_MD_NAME_MAX];
-        uct_md_h uct_md;
-        uct_ib_md_t *ib_md;
-        ucs_status_t status;
-        uint8_t gid_index;
-
-        /* no pkeys for Ethernet */
-        port_desc->have_pkey = 0;
+void test_uct_ib::initialize() {
```

Comment:
try to remove this function and "initialized" variable, the test should always be initialized by ::init() function

Reply:
done

`initialize` function and `bool initialized` variable was removed